### PR TITLE
D2 : routineNotCompleteOrServiceInProgress handling

### LINKED ISSUE #87: Two operations in rapid succession can confuse diag_l0_elm
If scantool initiates a command or request right after a previous request completed, the interface might still be listening for more responses from the vehicle when we send it the second operation. The interface will then send a STOPPED error, and the second operation will fail.

Repro cases: cleardtc followed by disconnect command on AW50-42 TCM (the ATPC command interrupts the stopDiagnosticSession transaction), or two cleardtcs in a row on AW50-42 TCM (the clearDiagnosticInformation request interrupts the readDiagnosticTroubleCodes transaction). cleardtc on AW50-42 sets the stage for this to happen on future operations because the >50ms response time causes the interface's adaptive timing algorithm to increase the response timeout on future requests.

See discussion in #84.

### PR DESCRIPTION
**Pull request by aaeegg**
D2 : If we get a routineNotCompleteOrServiceInProgress response, discard it and keep listening until we get the final response. This is necessary for requests that return a few routineNotCompleteOrServiceInProgresses before they complete, such as clearDiagnosticInformation on AW50-42.

This makes cleardtc work on AW50-42 (#84), but the interface will still increase its timeout after cleardtc on AW50-42 which can cause problems with other operations afterwards (does not fix #87).

### CODE REVIEW

File: scantool/diag_l2_d2.c [hunk 1]
```diff
@@ -147,14 +147,21 @@ static struct diag_msg *dl2p_d2_request(struct diag_l2_conn *d_l2_conn, struct d
...
-	if (rv < 0) {
-		*errval = rv;
-		return NULL;
-	}
-	if (rmsg == NULL) {
-		*errval = DIAG_ERR_NOMEM;
-	}
+	do {
+		if (rmsg != NULL)
+			diag_freemsg(rmsg);
```
Comment (fenugrec) [inline scantool/diag_l2_d2.c:152]: missing { }

File: scantool/diag_l2_d2.c [hunk 2]
```diff
@@ -147,14 +147,21 @@ static struct diag_msg *dl2p_d2_request(struct diag_l2_conn *d_l2_conn, struct d
...
+			*errval = rv;
+			return NULL;
+		}
+		if (rmsg == NULL) {
+			*errval = DIAG_ERR_NOMEM;
+			return NULL;
+		}
+		/* If we got routineNotCompleteOrServiceInProgress, loop until 
+		   the final response. */
+	} while(rmsg->len==3 && rmsg->data[0]==0x7e && rmsg->data[1]==msg->data[0] && rmsg->data[2]==0x23);
```
Comment (fenugrec) [inline scantool/diag_l2_d2.c:164]: space between `while` and `(`

**Review by fenugrec — CHANGES_REQUESTED** [review]
some minor style comments. Have you tried running uncrustify with our included .cfg ? I'm not enforcing it systematically (yet) but it should work acceptably.

Can you rework the commit message to follow the usual pattern (short first line summary, blank line, then explanation)

**Comment by aaeegg** [comment]
Amended with style changes.

**Comment by fenugrec** [comment]
Would it be possible to cover this in our 850 testsuite ? I think multiple responses should work with the carsim backend, it would be nice to have at least some coverage for this

**Comment by aaeegg** [comment]
Test added. I put the routineNotCompleteOrServiceInProgress test in dtc instead of cleardtc, because freediag won't take Yes/No responses from the ini file, so cleardtc would just sit forever in cli_basic_get_input waiting for keyboard input.

**Comment by fenugrec** [comment]
> Test added. I put the routineNotCompleteOrServiceInProgress test in dtc instead of cleardtc, because freediag won't take Yes/No responses from the ini file, so cleardtc would just sit forever in cli_basic_get_input waiting for keyboard input.

Sounds good. Thanks for taking the time to add those tests.

Rebased + merged at c546172c2d1df726f80ab19369ae0d48b943e0d5  !

**Pull request closed without merging**